### DISCUSSION TITLE
JSON Schema模式下，按钮和弹窗不在同一个组件时如何通讯？

### DISCUSSION THREAD

**Discussion by joyz0** [alibaba/formily#3967] [Q&A 2.x]
场景：按钮组件（MyButton）需要打开弹窗（具体是哪个弹窗由业务页面决定）并传递一些数据。

举例：
```js
// pageA
const schema = {
    type: 'void',
    'x-component': 'MyButton',
    properties: {
        modal: {
            type: 'void',
           'x-component': 'ModalA',
        }
    }
}

// pageB
const schema = {
    type: 'void',
    'x-component': 'MyButton',
    properties: {
        modal: {
            type: 'void',
           'x-component': 'ModalB',
        }
    }
}
```

在保证MyButton不存在任何业务逻辑的情况下（可以有约定式的逻辑），如何与外部通讯？
我目前能想到的方案是通过ReactContext，在MyButton维护两个状态（open和values），各自Modal各自取。但我想了解下有没有更契合Formily schema模式的方案？

**Comment by atom-set** [comment]
我也遇到类似的场景，我是通过props的方式传递的

**Reply by joyz0** [reply]
你是指`x-components-props`吗？

因为我是把scheme json当做静态声明使用的，所以在schema中访问不到React生命周期内的上下文。

**Reply by atom-set** [reply]
两种方式都行，可以通过x-components-props，也可以通过 react 的 props，看怎么方便接受

**Reply by joyz0** [reply]
还有两种思路：
1. observable和observer组合，比React Context好的一面是不要求双方在同一棵节点树中；
2. 使用form.notify()自定义事件来通讯。
我自己再试试，感受一下。

**Reply by atom-set** [reply]
如果组件层级深的可以，使用你说的这2种方式。

如果就是简单的父传子，我还是比较喜欢props方式

**Reply by joyz0** [reply]
我没有完全明白你的意思。MyButton组件和Modal组件不在一起，如何通过props传递？

**Reply by joyz0** [reply]
MyButton要打开哪个Modal是在不同业务使用的时候决定的。

**Reply by atom-set** [reply]
抱歉，我理解成了 MyButton 内包裹了 Modal，那可以使用你的2种方式。

我还有个疑问，MyButton 和 Modal 不在同一个文件内，MyButton 是怎么触发Modal 呢

**Reply by joyz0** [reply]
> MyButton 和 Modal 不在同一个文件内，MyButton 是怎么触发Modal 呢

我这个discussion的最初想问的问题就是你这个😁

**Reply by atom-set** [reply]
用 form.notify() 吧
